src/sentry/snuba/metrics/metric_extraction.py

```diff
@@ -0,0 +1,258 @@
+    # TODO(ogi): Support fields whose resolution returns a function
+    # "browser.name": "event.browser.name",
+    # "http.status_code": "event.http.status_code",
+    # "os.name": "event.os.name",
+}
+
+_SNUBA_TO_METRIC_AGGREGATES: Dict[str, Optional[MetricOperationType]] = {
+    "count": "sum",
+    "avg": "avg",
+    "quantile(0.50)": "p50",
+    "quantile(0.75)": "p75",
+    "quantile(0.95)": "p95",
+    "quantile(0.99)": "p99",
+    "max": "max",
+    # not supported yet
+    "percentile": None,
+    "failure_rate": None,
+    "apdex": None,
+}
+
+# TODO(ogi): support count_if
+_AGGREGATE_TO_METRIC_TYPE = {
+    "count": "c",
+    "avg": "d",
+    "quantile(0.50)": "d",
+    "quantile(0.75)": "d",
+    "quantile(0.95)": "d",
+    "quantile(0.99)": "d",
+    "max": "d",
+    # not supported yet
+    "percentile": None,
+    "failure_rate": None,
+    "apdex": None,
+}
+
+
+class ComparingRuleCondition(TypedDict):
+    """RuleCondition that compares a named field to a reference value."""
+
+    op: Literal["eq", "gt", "gte", "lt", "lte", "glob"]
+    name: str
+    value: Any
+
+
+class LogicalRuleCondition(TypedDict):
+    """RuleCondition that applies a logical operator to a sequence of conditions."""
+
+    op: Literal["and", "or"]
+    inner: Sequence[RuleCondition]
+
+
+class NotRuleCondition(TypedDict):
+    """RuleCondition that negates an inner condition."""
+
+    op: Literal["not"]
+    inner: RuleCondition
+
+
+class TagSpec(TypedDict):
+    """
+    Configuration for a tag to add to a metric.
+
+    Tags values can be static if defined through `value` or dynamically queried
+    from the payload if defined through `field`. These two options are mutually
+    exclusive, behavior is undefined if both are specified.
+    """
+
+    key: str
+    field: NotRequired[str]
+    value: NotRequired[str]
+    condition: NotRequired[RuleCondition]
+
+
+class MetricSpec(TypedDict):
+    """
+    Specification for a metric to extract from some data.
+
+    The metric type is given as part of the MRI (metric reference identifier)
+    which must follow the form: `<type>:<namespace>/<name>@<unit>`.
+
+    How the metric's value is obtained depends on the metric type:
+     - Counter metrics are a special case, since the default product counters do
+       not count any specific field but rather the occurrence of the event. As
+       such, there is no value expression, and the field is set to `None`.
+       Semantics of specifying remain undefined at this point.
+     - Distribution metrics require a numeric value.
+     - Set metrics require a string value, which is then emitted into the set as
+       unique value. Insertion of numbers and other types is undefined.
+    """
+
+    category: Literal["transaction"]
+    mri: str
+    field: NotRequired[Optional[str]]
+    condition: NotRequired[RuleCondition]
+    tags: NotRequired[Sequence[TagSpec]]
+
+
+def convert_alert_to_metric(snuba_query: SnubaQuery) -> Optional[MetricSpec]:
```

Comment:
While generating an on-demand metric from Discover queries is a more generic task, doing this sort of conversion seems like it's a relay config concern. Could we move this back to the Relay module?

Reply:
[bcc644e](https://github.com/getsentry/sentry/pull/52000/commits/bcc644e7084dcee4c6dd290a551b8db9bdba91f4)